cpp/velox/jni/JniWrapper.cc

```diff
@@ -154,6 +184,196 @@ JNIEXPORT void JNICALL Java_io_glutenproject_spark_sql_execution_datasources_vel
+  splitOptions.prefer_spill = prefer_spill;
+  splitOptions.buffered_write = true;
+  if (buffer_size > 0) {
+    splitOptions.buffer_size = buffer_size;
+  }
+  splitOptions.offheap_per_task = offheap_per_task;
+
+  if (num_sub_dirs > 0) {
+    splitOptions.num_sub_dirs = num_sub_dirs;
+  }
+
+  if (compression_type_jstr != NULL) {
+    auto compression_type_result = GetCompressionType(env, compression_type_jstr);
+    if (compression_type_result.status().ok()) {
+      splitOptions.compression_type = compression_type_result.MoveValueUnsafe();
+    }
+  }
+
+  auto data_file_c = env->GetStringUTFChars(data_file_jstr, JNI_FALSE);
+  splitOptions.data_file = std::string(data_file_c);
+  env->ReleaseStringUTFChars(data_file_jstr, data_file_c);
+
+  auto* allocator = reinterpret_cast<MemoryAllocator*>(allocator_id);
+  if (allocator == nullptr) {
+    gluten::JniThrow("Memory pool does not exist or has been closed");
+  }
+  splitOptions.memory_pool = gluten::AsWrappedArrowMemoryPool(allocator);
+
+  auto local_dirs = env->GetStringUTFChars(local_dirs_jstr, JNI_FALSE);
+  setenv("NATIVESQL_SPARK_LOCAL_DIRS", local_dirs, 1);
+  env->ReleaseStringUTFChars(local_dirs_jstr, local_dirs);
+
+  jclass cls = env->FindClass("java/lang/Thread");
+  jmethodID mid = env->GetStaticMethodID(cls, "currentThread", "()Ljava/lang/Thread;");
+  jobject thread = env->CallStaticObjectMethod(cls, mid);
+  if (thread == NULL) {
+    std::cout << "Thread.currentThread() return NULL" << std::endl;
+  } else {
+    jmethodID mid_getid = GetMethodIDOrError(env, cls, "getId", "()J");
+    jlong sid = env->CallLongMethod(thread, mid_getid);
+    splitOptions.thread_id = (int64_t)sid;
+  }
+
+  jclass tc_cls = env->FindClass("org/apache/spark/TaskContext");
+  jmethodID get_tc_mid = env->GetStaticMethodID(tc_cls, "get", "()Lorg/apache/spark/TaskContext;");
+  jobject tc_obj = env->CallStaticObjectMethod(tc_cls, get_tc_mid);
+  if (tc_obj == NULL) {
+    std::cout << "TaskContext.get() return NULL" << std::endl;
+  } else {
+    jmethodID get_tsk_attmpt_mid = GetMethodIDOrError(env, tc_cls, "taskAttemptId", "()J");
+    jlong attmpt_id = env->CallLongMethod(tc_obj, get_tsk_attmpt_mid);
+    splitOptions.task_attempt_id = (int64_t)attmpt_id;
+  }
+  splitOptions.batch_compress_threshold = batch_compress_threshold;
+
+  std::shared_ptr<ColumnarBatch> batch = ColumnBatchHolder::lookup(firstBatchHandle);
+  jlong splitterId;
+  if (batch->GetType() == "velox") {
+    auto veloxSplitter = gluten::JniGetOrThrow(
+        VeloxSplitter::Make(partitioning_name, num_partitions, std::move(splitOptions)),
+        "Failed create native velox shuffle splitter");
+    splitterId = shuffle_splitter_holder_.Insert(veloxSplitter);
+  } else {
+    auto splitter = gluten::JniGetOrThrow(
+        Splitter::Make(partitioning_name, num_partitions, std::move(splitOptions)),
+        "Failed create native shuffle splitter");
+    splitterId = shuffle_splitter_holder_.Insert(splitter);
+  }
```

Comment:
So if `firstBatchHandle` is just for picking spliter for the batch type, can we possibly do this per batch, to avoid specifically check the first one?

Reply:
Because not only `split` function need splitter, the function `nativeSpill` and `stop` also need the splitter, we can not get bathc type there

Reply:
Okay. Not a big deal and we may optimize on this later.